man/cat1/xray.1



XRAY(1)                  TIPSY COMMANDS                   XRAY(1)



NAME
     xray - make xray maps in different energy bands and write to
     FITS files

SYNOPSIS
ALIASES
DESCRIPTION
SEE ALSO












































Katz and Quinn ReleasLeas2t.0change: 22 MARCH 1994                  1








code/ChangeLog
Wed Jan 11 13:09:36 1995  Tom Quinn  <trq@hermes.astro.washington.edu>

	* Makefile.in (SRCS): added alloca.c so it will be included in the
	distribution.

Wed Dec 28 13:42:47 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* readmark.c: close file after it is read.

	* main.c: RedrawViewport(): make can_size the maximum of
	can_height and can_width.

	* canvas_resize_proc.c: Same as above.

Mon Dec  5 09:39:44 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* xray_lum_load.c, star_lum.c, star_lum_redshift.c: Changed
	argument to my_gets() to blank string.

Fri Dec  2 08:21:50 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* Makefile.in (READLINE_SRC): delete config.h.
	(tipsy): added dependence on ALLOCA.

	* fits.c: Added test for endianness.

Thu Dec  1 15:14:58 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* vista.c: fixed initialization of quantity[].

	* find_shape.c: Forced inertia_cm to be symmetric.  This avoid
	problems with jacobi().

Tue Nov  1 11:26:56 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* Makefile.in : fits_xray.c is now gone.  fits() is used instead.
	defs.h, xray.c, vista.c, fits.c and arguments.c are also changed.

Mon Oct 31 09:43:29 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* main.c: RedrawColorbar(): fixed off-by-one drawing errors.

Wed Oct 12 14:27:16 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* Created version 2.1.2.

	* absorb.c: New function for calculating aborbtion spectra.

Tue Sep  6 16:06:02 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readline/: Replaced with readline-2.0 from prep.ai.mit.edu.

	* my_gets.c: Changed cast to match new external from readline/.

Fri Jul 29 10:25:49 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* project.c: Ensure that the axes projections do not overflow.

Fri Jul 15 17:28:15 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* profile.c: Check to see if fopen() fails.

Wed Jul 13 09:46:18 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* defs.h: Decreased INTMAX to 2^29-1.  (One bit for sign, one bit
	to do INTMAX - (-INTMAX), and one bit for zoom out.

Tue Jul 12 15:55:17 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* canvas_resize_proc.c: Call plot_sub() and draw_label() for
	servers that automatically do backing store and don't generate an
	expose event.

	* main.c: ZoomIn(), ZoomOut(), Pan(), SetBox(): Ignore button
	presses in inactive windows.
	
	* main.c, window.c: Changed ResizeRequest to ConfigureNotify.

Sat Jun  4 13:12:06 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* fits.c, and fits_xray.c: Changed write() to fwrite() for
	portability.

Thu May  5 10:53:45 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* project.c: Increase clipping range of velocities to -2*INTMAX <
	v < 2*INTMAX.
	* defs.h: Change INTMAX to 2^30-1 to accommodate the above
	clipping.

Tue Apr 19 18:00:14 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* comm.h: NSK: added pcenter and writeboxrot commands.

Mon Apr 18 14:14:45 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* main.c: Add "unzoom" button.

Fri Apr 15 16:01:56 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Fixed bug with color of marked particles.
	Deleted extra conversion to screen coordinates.
	
	* maxpos.c: Make the margins a little smaller.

Thu Mar 31 14:03:11 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Changed XDrawArc to XFillArc for vector draws.

Fri Mar  4 18:09:25 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* Added Neal's changes for read_vector() and "master".

Mon Feb 28 11:51:39 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* openascii.c: Added code from Peter Teuben to uncompress files.

Thu Feb 24 13:21:50 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* Changed current_plot logical to current_project and
	current_color.

Tue Feb 22 11:24:05 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readline: Updated to readline from gdb-1.12

Thu Feb 17 11:46:07 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Implimented a software "z-buffer".  Only draw
	particles with the highest particle color.

Wed Feb 16 13:21:35 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Merged plot_all and plot_vel into one routine.  A
	new routine all_color() is created.  Changes were also needed in
	the view_*() routines.

Sat Nov 13 11:35:01 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readmark.c: Fixed typo for increasing ndark.

Wed Nov  3 14:35:56 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Deleted skip of XClearWindow on first time through.

Thu Oct 28 10:23:50 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* hard.c: Added movie parameter.

Tue Oct 26 11:29:49 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* alloca.c: New file in case system doesn't provide alloca().

	* vista.c: Included <stdlib.h> for malloc() definition.

Mon Oct 25 19:16:52 1993  Tom Quinn  (trq@virgo)

	* main.c: Attempt to find a depth 8 PseudoColor visual.

Thu Oct 21 08:56:32 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* main.c: Set backgrounds of gc_color[] to black.

Tue Oct 19 10:56:36 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* defs.h, main.h, make_macro.c, read_macro.c, write_macro.c
	run_macro.c: re-vamped macro storage using linked lists.

Mon Oct 18 15:11:34 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* draworient.c: Added cast from Arg to long.

Fri Oct 15 10:04:49 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readmark.c: Added a check for index < 1.

	* plot_sub.c: Added a check for binary_loaded.
	Added check for boxes_loaded[0].

Wed Oct  6 09:13:49 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* zoom_window.c: Added zoom command.  Also changed comm.h and
	arguments.c.

Tue Oct  5 17:27:43 1993  Tom Quinn  (trq@hermes.astro.washington.edu

code/xray.c
/* $Header$
 * $Log$
 * Revision 1.5  2007/01/14 19:43:22  trq
 * 1) load redshift information if needed.
 * 2) add warning about floating point overflow of temp_array.
 *
 * Revision 1.4  2000/01/12 22:55:29  nsk
 * Fixed bugs in cooling routines, added cooling damping,
 * fixed bugs in starformation,  regularized dependencies.
 *
 * Revision 1.3  1998/02/18  22:54:56  trq
 * Correct bounding on temperature.
 *
 * Revision 1.2  1997/12/06  22:00:57  trq
 * Updated xray stuff to use spline emmission files.
 *
 * Revision 1.1.1.1  1995/01/10  22:57:37  trq
 * Import to CVS
 *
 * Revision 2.4  1994/11/01  19:20:04  nsk
 * allowed size of image to change nad call to more general fits.c
 *
 * Revision 2.3  1994/10/12  21:26:39  trq
 * Cleaned up mallocs.
 *
 * Revision 2.2  1994/04/20  08:46:30  trq
 * Added title variable.
 *
 * Revision 2.0  93/10/19  13:24:08  trq
 * *** empty log message ***
 * 
 */

#include "defs.h"
#include "fdefs.h"
#include <stdlib.h>
#include <values.h>
#include <assert.h>
#include "xray.h"

void
xray(job)
    char *job;
{
    char command[MAXCOMM] ;
    double size_pixel ;
    double size_pixel_2 ;
    double xmin ;
    double ymin ;
    struct gas_particle *gp ;
    double pixel_pos[2] ;
    double part_pos[2] ;
    int i,j ;
    int kx,ky ;
    int kx_min,kx_max ;
    int ky_min,ky_max ;
    double ***density;
    double radius2 ;
    double distance_dim2() ;
    double hsmooth ;
    double thsmooth ;
    double hsmth2pi ;
    double kernel ;
    char name[50] ;
    char image_name[50] ;
    float **image;
    float **temp_image;
    double pixel ;
    double low ;
    double high ;
    double distnorm ;
    double drw ;
    int iwsm ;
    int xray_lum_load() ;
    double *lum_xray ;
    int band ;
    double *band_lum ;
    int xray_size ;

    double total_lum ; 

    if (sscanf(job,"%s %lf %lf %s %d",command,&low,&high,name,&xray_size) == 
	    5) {
	if(!ikernel_loaded){
	    ikernel_load() ;
	}
	if (!redshift_loaded ){
	    load_redshift();
	    }
	if (current_project){
	    if(!xray_loaded){
		if(xray_lum_load() != 0){
		    return ;
		}
	    }
	    if (!cool_loaded ){
	      printf("loading cool %d \n",cool_loaded);
	       load_cool() ;
		}
	    
	  density = (double ***)malloc(number_bands*sizeof(*density));
	  if(density == NULL)
	    {
	      printf("<sorry, no memory for image, %s>\n",title) ;
	      return ;
	    }
	  *density = (double **)malloc(number_bands*xray_size*sizeof(**density));
	  if(*density == NULL)
	    {
	      free(density);
	      printf("<sorry, no memory for image, %s>\n",title) ;
	      return ;
	    }
	  image = (float **)malloc(xray_size*sizeof(*image));
	  if(image == NULL)
	    {
	      free(density);
	      free(*density);
	      printf("<sorry, no memory for image, %s>\n",title) ;
	      return ;
	    }
	  temp_image = (float **)malloc(xray_size*sizeof(*temp_image));
	  if(temp_image == NULL)
	    {
	      free(density);
	      free(*density);
	      free(image);
	      printf("<sorry, no memory for image, %s>\n",title) ;
	      return ;
	    }
	  **density = (double *)malloc(number_bands*xray_size*xray_size*
		sizeof(***density));
	  if(**density == NULL)
	    {
	      free(density);
	      free(*density);
	      free(image);
	      free(temp_image);
	      printf("<sorry, no memory for image, %s>\n",title) ;
	      return ;
	    }
	  *image = (float *)malloc(xray_size*xray_size*sizeof(**image));
	  if(*image == NULL)
	    {
	      free(density);
	      free(image);
	      free(temp_image);
	      free(*density);
	      free(**density);
	      printf("<sorry, no memory for image, %s>\n",title) ;
	      return ;
	    }
	  *temp_image = (float *)malloc(xray_size*xray_size*sizeof(**temp_image));
	  if(*temp_image == NULL)
	    {
	      free(density);
	      free(image);
	      free(temp_image);
	      free(*image);
	      free(*density);
	      free(**density);
	      printf("<sorry, no memory for image, %s>\n",title) ;
	      return ;
	    }
	  for(j = 1; j < number_bands; j++)
	      density[j] = &density[0][j*xray_size];
	  
	  for(i = 0; i < xray_size; i++)
	    {
	      for(j = 0; j < number_bands; j++)
		density[j][i] = &density[0][0][j*xray_size*xray_size + 
			i*xray_size];
	      image[i] = &image[0][i*xray_size];
	      temp_image[i] = &temp_image[0][i*xray_size];
	    }
	    size_pixel = (dv2_x - dv1_x) / scaling / (double)xray_size ;
	    size_pixel_2 = size_pixel*size_pixel ;
	    xmin = dv1_x / scaling ;
	    ymin = dv1_y / scaling ;
	    band_lum = (double *)malloc((number_bands)*sizeof(*band_lum)) ;
	    for(band = 0; band < number_bands; band++){
		for(kx = 0; kx < xray_size; kx++){
		    for(ky = 0; ky < xray_size; ky++){
		    density[band][kx][ky] = 0.0 ;
		    }
		}
	    }
	    for(kx = 0; kx < xray_size; kx++){
		for(ky = 0; ky < xray_size; ky++){
		    temp_image[kx][ky] = 0.0 ;
		}
	    }
	    for (i = 0 ;i < boxlist[active_box].ngas;i++) {
		gp = boxlist[active_box].gp[i] ;
		if(log10((double) gp->temp) > xray_min_temp){
		    for(band = 0,lum_xray = band_lum; band < number_bands;
			    band++,lum_xray++){
			 *lum_xray = gp->mass*calc_xemiss((double) gp->temp,
						 (double) gp->rho, band);
			
			if(*lum_xray < 0.){
			    printf("<sorry, negative luminosity %s\n",title) ;
			    return ;
			}
			
		    }
		    hsmooth = gp->hsmooth ;
		    for (part_pos[0] = part_pos[1] = 0.0,j = 0 ;
			 j < header.ndim ;j++) {
			part_pos[0] += rot_matrix[0][j] * (gp->pos[j] -
				    boxes[active_box].center[j]) ;
			part_pos[1] += rot_matrix[1][j] * (gp->pos[j] -
				    boxes[active_box].center[j]) ;
		    }
		    if(hsmooth > size_pixel) {
			thsmooth = 2. * hsmooth ;
			distnorm = 1. / (hsmooth * hsmooth) ;
			hsmth2pi = distnorm / PI ;
			kx_min = max(0,(int)((part_pos[0]-thsmooth-xmin)/
				size_pixel + .499999)) ;
			kx_max = min(xray_size-1,(int)((part_pos[0]+thsmooth-
				xmin)/
				size_pixel + .499999)) ;
			ky_min = max(0,(int)((part_pos[1]-thsmooth-ymin)/
			    size_pixel + .499999)) ;
			ky_max = min(xray_size-1,(int)((part_pos[1]+thsmooth-
				ymin)/
				size_pixel + .499999)) ;
			for(kx = kx_min; kx < kx_max + 1; kx++){
			    for(ky = ky_min; ky < ky_max + 1; ky++){
				pixel_pos[0] = xmin + (kx + .5) * size_pixel ;
				pixel_pos[1] = ymin + (ky + .5) * size_pixel ;
				radius2 = distance_dim2(pixel_pos,part_pos) *
					distnorm ;
				if(radius2 < 4.){
				    radius2 *= deldr2i ;
				    iwsm = (int)radius2 ;
				    drw = radius2 - (double)iwsm ;
				    kernel = ((1. - drw) * iwsmooth[iwsm] +
					    drw * iwsmooth[1 + iwsm]) * hsmth2pi ;
				}
				else{
				    kernel = 0. ;
				}
				kernel *= size_pixel_2 ;
				if(kernel != 0.){
				    for(band = 0; band < number_bands; band++){
					double tTemp
					    = kernel*band_lum[band]*gp->temp;
					if(tTemp > FLT_MAX) {
					    fprintf(stderr,
						    "Warning: %g overflows temperature array\n",
						    tTemp);
					    tTemp = FLT_MAX/10.0;
					    }
					
					density[band][kx][ky] += (kernel*
					    band_lum[band]) ;
					temp_image[kx][ky] += tTemp;
				    }
				}
			    }
			}
		    }
		    else {
			kx = (int)((part_pos[0]-xmin)/size_pixel+ 0.499999) ;
			ky = (int)((part_pos[1]-ymin)/size_pixel+ 0.499999) ;
			if(kx >= 0 && kx < xray_size && ky >= 0 && ky
			   < xray_size){
			    for(band = 0; band < number_bands; band++){
				double tTemp = band_lum[band]*gp->temp;
				if(tTemp > FLT_MAX) {
				    fprintf(stderr, "Warning: %g overflows temperature array\n",
					    tTemp);
				    tTemp = FLT_MAX/10.0;
				    }
				density[band][kx][ky] += band_lum[band];
				temp_image[kx][ky]
					+= tTemp;
			    }
			}
		    }
		}
	    }
	    for(band = 0; band < number_bands; band++){

	      total_lum = 0.0 ;

		for(i = 0; i < xray_size; i++){
		    for(j = 0; j < xray_size; j++){
			if(density[band][i][j] > 0. ){

			  total_lum += (double)(density[band][i][j]) ;

			    pixel = log10((double)(density[band][i][j]));
			}
			else{
			    pixel = low ;
			}
			if(pixel > high)pixel = high ;
			if(pixel < low)pixel = low ;
			image[i][j] = (float)pixel ;
		    }
		}
		printf(" Band Number: %d flux = %f \n",band,log10(total_lum) );
		sprintf(image_name,"%s%03d.fits",name,band) ;
		fits(image,xray_size,xray_size,0.,
			0.,1.,1.,low,high,image_name) ;
	    }
	/* sum up the luminosity in all bands. */
	    for(band = 1; band < number_bands; band++){
		for(i = 0; i < xray_size; i++){
		    for(j = 0; j < xray_size; j++){
			density[0][i][j] += density[band][i][j];
			}
		    }
		}
	    for(i = 0; i < xray_size; i++){
		for(j = 0; j < xray_size; j++){
		    if(density[0][i][j] > 0. ){
			pixel = log10((double)(density[0][i][j]));
		    }
		    else{
			pixel = low ;
		    }
		    if(pixel > high)pixel = high ;
		    if(pixel < low)pixel = low ;
		    image[i][j] = (float)pixel ;
		    temp_image[i][j] /= density[0][i][j];
#define LOW_TEMP 0.0
#define HIGH_TEMP 20.0
		    if(temp_image[i][j] > 0. ){
			pixel = log10((double)(temp_image[i][j]));
		    }
		    else{
			pixel = LOW_TEMP ;
		    }
		    if(pixel > HIGH_TEMP)pixel = HIGH_TEMP ;
		    if(pixel < LOW_TEMP)pixel = LOW_TEMP ;
		    temp_image[i][j] = (float) pixel;
		}
	    }
	    sprintf(image_name,"%stot.fits",name) ;
	    fits(image,xray_size,xray_size,0.,
		    0.,1.,1.,low,high,image_name) ;
	    sprintf(image_name,"%stemp.fits",name) ;
	    fits(temp_image,xray_size,xray_size,0.,
		    0.,1.,1.,LOW_TEMP,HIGH_TEMP,image_name) ;
	    
	}
	else {
	    printf("<plot does not represent the present internal state, %s>",title) ;
	}
      free(**density);
      free(*density);
      free(density);
      free(*image);
      free(image);
      free(*temp_image);
      free(temp_image);
    }
    else {
	input_error(command) ;
    }
}
